File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Server~/src/tools/createPrefabTool.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import * as z from 'zod';
import { Logger } from '../utils/logger.js';

// Constants for the tool
const toolName = 'create_prefab';
const toolDescription = 'Creates a prefab with optional MonoBehaviour script and serialized field values';

// Parameter schema for the tool
const paramsSchema = z.object({
  componentName: z.string().optional().describe('The name of the MonoBehaviour Component to add to the prefab (optional)'),
  prefabName: z.string().describe('The name of the prefab to create'),
  fieldValues: z.record(z.any()).optional().describe('Optional JSON object of serialized field values to apply to the prefab')
});

/**
 * Creates and registers the CreatePrefab tool with the MCP server
 * 
 * @param server The MCP server to register the tool with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerCreatePrefabTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);
  
  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

/**
 * Handler function for the CreatePrefab tool
 * 
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The validated parameters for the tool
 * @returns A promise that resolves to the tool execution result
 * @throws McpUnityError if validation fails or the request to Unity fails
 */
async function toolHandler(mcpUnity: McpUnity, params: any) {
  if (!params.prefabName) {
    throw new McpUnityError(
      ErrorType.VALIDATION,
      "'prefabName' must be provided"
    );
  }
  
  const response = await mcpUnity.sendRequest({
    method: toolName,
    params
  });
  
  if (!response.success) {
    throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      response.message || `Failed to create prefab`
    );
  }
  
  return {
    content: [{
      type: response.type,
      text: response.message || `Successfully created prefab`
    }],
    // Include the prefab path in the result for programmatic access
    data: {
      prefabPath: response.prefabPath
    }
  };
}

File: Workpiece/Summoner/LocalPackages/com.gamelovers.mcp-unity/Editor/Tools/CreatePrefabTool.cs
using System;
using UnityEngine;
using UnityEditor;
using Newtonsoft.Json.Linq;
using McpUnity.Unity;
using McpUnity.Utils;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for creating prefabs with optional MonoBehaviour scripts
    /// </summary>
    public class CreatePrefabTool : McpToolBase
    {
        public CreatePrefabTool()
        {
            Name = "create_prefab";
            Description = "Creates a prefab with optional MonoBehaviour script and serialized field values";
        }
        
        /// <summary>
        /// Execute the CreatePrefab tool with the provided parameters
        /// </summary>
        /// <param name="parameters">Tool parameters as a JObject</param>
        public override JObject Execute(JObject parameters)
        {
            // Extract parameters
            string componentName = parameters["componentName"]?.ToObject<string>();
            string prefabName = parameters["prefabName"]?.ToObject<string>();
            JObject fieldValues = parameters["fieldValues"]?.ToObject<JObject>();
            
            // Validate required parameters
            if (string.IsNullOrEmpty(prefabName))
            {
                return McpUnitySocketHandler.CreateErrorResponse(
                    "Required parameter 'prefabName' not provided", 
                    "validation_error"
                );
            }
            
            // Create a temporary GameObject
            GameObject tempObject = new GameObject(prefabName);

            // Add component if provided
            if (!string.IsNullOrEmpty(componentName))
            {
                try
                {
                    // Add component
                    Component component = AddComponent(tempObject, componentName);
            
                    // Apply field values if provided and component exists
                    ApplyFieldValues(fieldValues, component);
                }
                catch (Exception)
                {
                    return McpUnitySocketHandler.CreateErrorResponse(
                        $"Failed to add component '{componentName}' to GameObject", 
                        "component_error"
                    );
                }
            }
            
            // For safety, we'll create a unique name if prefab already exists
            int counter = 1;
            string prefabPath = $"{prefabName}.prefab";
            while (AssetDatabase.AssetPathToGUID(prefabPath) != "")
            {
                prefabPath = $"{prefabName}_{counter}.prefab";
                counter++;
            }
            
            // Create the prefab
            bool success = false;
            PrefabUtility.SaveAsPrefabAsset(tempObject, prefabPath, out success);
            
            // Clean up temporary object
            UnityEngine.Object.DestroyImmediate(tempObject);
            
            // Refresh the asset database
            AssetDatabase.Refresh();
            
            // Log the action
            McpLogger.LogInfo($"Created prefab '{prefabName}' at path '{prefabPath}' from script '{componentName}'");

            string message = success ? $"Successfully created prefab '{prefabName}' at path '{prefabPath}'" : $"Failed to create prefab '{prefabName}' at path '{prefabPath}'";
            
            // Create the response
            return new JObject
            {
                ["success"] = success,
                ["type"] = "text",
                ["message"] = message,
                ["prefabPath"] = prefabPath
            };
        }

        private Component AddComponent(GameObject gameObject, string componentName)
        {
            // Find the script type
            Type scriptType = Type.GetType($"{componentName}, Assembly-CSharp");
            if (scriptType == null)
            {
                // Try with just the class name
                scriptType = Type.GetType(componentName);
            }
                
            if (scriptType == null)
            {
                // Try to find the type using AppDomain
                foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
                {
                    scriptType = assembly.GetType(componentName);
                    if (scriptType != null)
                        break;
                }
            }
                
            // Throw an error if the type was not found
            if (scriptType == null)
            {
                return null;
            }
                
            // Check if the type is a MonoBehaviour
            if (!typeof(MonoBehaviour).IsAssignableFrom(scriptType))
            {
                return null;
            }
            
            return gameObject.AddComponent(scriptType);
        }

        private void ApplyFieldValues(JObject fieldValues, Component component)
        {
            // Apply field values if provided and component exists
            if (fieldValues == null || fieldValues.Count == 0)
            {
                return;
            }
            
            Undo.RecordObject(component, "Set field values");
                
            foreach (var property in fieldValues.Properties())
            {
                // Get the field/property info
                var fieldInfo = component.GetType().GetField(property.Name, 
                    System.Reflection.BindingFlags.Public | 
                    System.Reflection.BindingFlags.NonPublic | 
                    System.Reflection.BindingFlags.Instance);
                            
                if (fieldInfo != null)
                {
                    // Set field value
                    object value = property.Value.ToObject(fieldInfo.FieldType);
                    fieldInfo.SetValue(component, value);
                }
                else
                {
                    // Try property
                    var propInfo = component.GetType().GetProperty(property.Name, 
                        System.Reflection.BindingFlags.Public | 
                        System.Reflection.BindingFlags.NonPublic | 
                        System.Reflection.BindingFlags.Instance);
                                
                    if (propInfo != null && propInfo.CanWrite)
                    {
                        object value = property.Value.ToObject(propInfo.PropertyType);
                        propInfo.SetValue(component, value);
                    }
                }
            }
        }
    }
}
